packages/contracts-bedrock/test/Initializable.t.sol

```diff
@@ -2,59 +2,246 @@
+            bytes32 initSlotVal = vm.load(_contract.target, bytes32(vm.parseUint(_contract.initializedSlot.slot)));
+
+            // Pull out the 8-bit `_initialized` flag from the storage slot. The offset in forge artifacts is
+            // relative to the least-significant bit and signifies the *byte offset*, so we need to shift the
+            // value to the right by the offset * 8 and then mask out the low-order byte to retrieve the flag.
+            uint8 init = uint8((uint256(initSlotVal) >> (initSlotOffset * 8)) & 0xFF);
+            assertEq(init, 3);
+
+            // Then, attempt to re-initialize the contract. This should fail.
+            (bool success, bytes memory returnData) = _contract.target.call(_contract.initCalldata);
+            assertFalse(success);
+            assertEq(_extractErrorString(returnData), "Initializable: contract is already initialized");
+        }
+    }
+
+    /// @dev Pulls the `_initialized` storage slot information from the Forge artifacts for a given contract.
+    function _getInitializedSlot(string memory _contractName) internal returns (StorageSlot memory slot_) {
+        string memory storageLayout = getStorageLayout(_contractName);
+
+        string[] memory command = new string[](3);
+        command[0] = Executables.bash;
+        command[1] = "-c";
+        command[2] = string.concat(
+            Executables.echo,
+            " '",
+            storageLayout,
+            "'",
+            " | ",
+            Executables.jq,
+            " '.storage[] | select(.label == \"_initialized\" and .type == \"t_uint8\")'"
+        );
+        bytes memory rawSlot = vm.parseJson(string(vm.ffi(command)));
+        slot_ = abi.decode(rawSlot, (StorageSlot));
+    }
+
+    /// @dev Returns the number of contracts that are `Initializable` in `src/L1`.
+    function _getNumL1Initializable() internal returns (uint256 numContracts_) {
+        string[] memory command = new string[](3);
+        command[0] = Executables.bash;
+        command[1] = "-c";
+        command[2] = string.concat(
+            Executables.find,
+            " src/L1 -type f -exec basename {} \\;",
+            " | ",
+            Executables.sed,
+            " 's/\\.[^.]*$//'",
+            " | ",
+            Executables.jq,
+            " -R -s 'split(\"\n\")[:-1]'"
+        );
+        string[] memory contractNames = abi.decode(vm.parseJson(string(vm.ffi(command))), (string[]));
+
+        for (uint256 i; i < contractNames.length; i++) {
+            string memory contractName = contractNames[i];
+            string memory contractAbi = getAbi(contractName);
+
+            // Query the contract's ABI for an `initialize()` function.
+            command[2] = string.concat(
+                Executables.echo,
+                " '",
+                contractAbi,
+                "'",
+                " | ",
+                Executables.jq,
+                " '.[] | select(.name == \"initialize\" and .type == \"function\")'"
```

Comment:
They are, however there's currently not a clean way for us to enforce this. OZ's `Initializable` abstract contract only includes modifiers, and in order to detect initializer functions with a separate name, you'd need to crawl through the AST.

Reply:
We can crawl for initializer decorated functions in the AST, but solc's AST ordering is non-deterministic, making this test w/ the AST crawler ripe for flaking in the future.

```
jq '.ast.nodes[] | select(.nodeType == "ContractDefinition")' | jq '.nodes[] | select(.nodeType == "FunctionDefinition" and .name == "initialize" and .modifiers[0].modifierName.name == "initializer")'
```

Specifically:
1. There can be multiple `ContractDefinition` nodes, i.e. if the file has multiple `contract XYZ { /* ... */ }` declarations.
1. The function can have multiple modifiers within it, making indexing the array of modifiers non-deterministic w/o ripping it out into its own check.

We have to manually craft the calldata for each of these `initialize` calls anyways, since inferring the types of the `initialize(..)` func's args wouldn't be typesafe w/o it, so I don't see this as an issue.

Reply:
Got it, as long as they are all called `initialize` (and from a quick search this seems to be the case) then the current approach seems good to me. This does imply we now have a restriction that all initializer methods must be called `initialize`, to ensure they get picked up by this test—perhaps that would be worth adding as a sempgrep rule?

Reply:
Yeah, could be a good check for Semgrep. Shouldn't be too difficult to write a regex for that 😄

Reply:
Sounds good, we can resolve this once that's added or tracked in a separate issue—either is ok with me, so deferring to you and feel free to resolve this comment

Reply:
https://github.com/ethereum-optimism/client-pod/issues/236